### PR TITLE
Add cvector_metadata_t

### PR DESCRIPTION
**Pull request by dm9pZCAq**
Use struct for more convenient way to access cvector metadata

(this would help to add more fields to cvector metadata for example like in #45)

### CODE REVIEW

**Comment by eteran** [comment]
I'll take a look and confirm correctness. I've been wanting to use a struct, but in the past, some nuances felt a bit fiddly to get right. Thanks!

File: cvector.h [hunk 1]
```diff
@@ -301,22 +306,22 @@ typedef void (*cvector_elem_destructor_t)(void *elem);
...
-            void *cv_p__ = cvector_clib_malloc(cv_sz__);                                                          \
-            assert(cv_p__);                                                                                       \
-            (vec) = cvector_base_to_vec(cv_p__);                                                                  \
-            cvector_set_size((vec), 0);                                                                           \
-            cvector_set_elem_destructor((vec), NULL);                                                             \
-        }                                                                                                         \
-        cvector_set_capacity((vec), (count));                                                                     \
+#define cvector_grow(vec, count)                                                      \
+    do {                                                                              \
+        const size_t cv_sz__ = (count) * sizeof(*(vec)) + sizeof(cvector_metadata_t); \
```
Comment (eteran) [inline cvector.h:311]: Honestly, this bit is really the only thing I have concerns about off hand. Alignment stuff can get really weird. So we wanna make sure that it **always** ends up with the user pointer landing in the right spot relative to the meta-data, especially given that the meta-data is basically the size of 3 pointers, and 3 is an size multiple for alignment concerns.

Comment (dm9pZCAq) [reply cvector.h:311]: I checked it myself and everything works and GitHub action test also passes.
About alignment: we can use packed struct ([`__attribute__((__packed__))`](https://gcc.gnu.org/onlinedocs/gcc/Common-Type-Attributes.html))

Comment (dm9pZCAq) [reply cvector.h:311]: here is memory layout of unpacked struct:
pic


<details><summary>code</summary>

```c
#include <stdio.h>
#include <inttypes.h>
#include <stdint.h>

#define CVECTOR_LOGARITHMIC_GROWTH
#include "cvector.h"

void
print_hex(const char *title, uint8_t *data, size_t size)
{
	printf("\033[36m%s\033[0m:\n", title);
	printf("\033[33m%p\033[0m: \033[32m", data);
	for (size_t i = 0; i < size; i++) {
		printf(!!data[i] ? "\033[1m0x%02x\033[0;32m " : "0x%02x ",
		       data[i]);
	}
	printf("\033[0m\n");
}

void
dump_metadata(cvector_metadata_t *metadata)
{
	print_hex("metadata", (void *)metadata, sizeof(*metadata));
	print_hex("metadata.size",
		  (void *)&metadata->size,
		  sizeof(metadata->size));
	print_hex("metadata.capacity",
		  (void *)&metadata->capacity,
		  sizeof(metadata->capacity));
	print_hex("metadata.elem_destructor",
		  (void *)&metadata->elem_destructor,
		  sizeof(metadata->elem_destructor));
	putchar('\n');
}

void
destructor(void *x)
{
	(void)x;
}

int
main(void)
{
	cvector_vector_type(uint8_t) vec = NULL;

	cvector_push_back(vec, 1);
	fprintf(stderr, "add 1\n");
	dump_metadata(cvector_vec_to_base(vec));

	cvector_set_elem_destructor(vec, destructor);

	cvector_push_back(vec, 2);
	fprintf(stderr, "add 2\n");
	dump_metadata(cvector_vec_to_base(vec));

	cvector_push_back(vec, 3);
	fprintf(stderr, "add 3\n");
	dump_metadata(cvector_vec_to_base(vec));

	fprintf(stderr, "metadata size: %zu\n", sizeof(cvector_metadata_t));
	fprintf(stderr, "start: %p\n", cvector_vec_to_base(vec));
	fprintf(stderr, "userptr: %p\n", vec);
	fprintf(stderr,
		"diff: %tu\n",
		(void *)vec - (void *)cvector_vec_to_base(vec));

	return 0;
}
```

</details>

Comment (eteran) [reply cvector.h:311]: Away from my computer tonight but this looks VERY promising, thanks!

Comment (dm9pZCAq) [reply cvector.h:311]: Oh, actually attached image is not correct

Here is correct image for `metadata_in_struct` branch:
pic

-----

Alignment ensures that data is stored at addresses that are a multiple of a certain value, which is determined by the architecture and can be the size of a processor register. On x86-64 architecture, the register size is 8 bytes, so data is aligned to 8-byte boundaries.

Here is picture of:
```c
typedef struct cvector_metadata_t {
    char x;  /* 1-byte field */
    size_t size, capacity;
    cvector_elem_destructor_t elem_destructor;
} cvector_metadata_t;
```
pic

You can play with it yourself by changing type of `x` field of `cvector_metadata_t` in `cvector.h`:
<details><summary>code</summary>

```c
#include <stdio.h>
#include <inttypes.h>
#include <stdint.h>

#define CVECTOR_LOGARITHMIC_GROWTH
#include "cvector.h"

void
print_hex(const char *title, uint8_t *data, size_t size)
{
	printf("\033[36m%s\033[0m:\n", title);
	printf("\033[33m%p\033[0m: \033[32m", data);
	for (size_t i = 0; i < size; i++) {
		printf(!!data[i] ? "\033[1m0x%02x\033[0;32m " : "0x%02x ",
		       data[i]);
	}
	printf("\033[0m\n");
}

void
dump_metadata(cvector_metadata_t *metadata)
{
	print_hex("metadata", (void *)metadata, sizeof(*metadata));
	print_hex("metadata.x",
		  (void *)&metadata->x,
		  sizeof(metadata->x));
	print_hex("metadata.size",
		  (void *)&metadata->size,
		  sizeof(metadata->size));
	print_hex("metadata.capacity",
		  (void *)&metadata->capacity,
		  sizeof(metadata->capacity));
	print_hex("metadata.elem_destructor",
		  (void *)&metadata->elem_destructor,
		  sizeof(metadata->elem_destructor));
	putchar('\n');
}

void
destructor(void *x)
{
	(void)x;
}

int
main(void)
{
	cvector_vector_type(uint8_t) vec = NULL;

	cvector_push_back(vec, 1);
	fprintf(stderr, "add 1\n");
	dump_metadata(cvector_vec_to_base(vec));

	cvector_set_elem_destructor(vec, destructor);

	cvector_push_back(vec, 2);
	fprintf(stderr, "add 2\n");
	dump_metadata(cvector_vec_to_base(vec));

	cvector_push_back(vec, 3);
	fprintf(stderr, "add 3\n");
	dump_metadata(cvector_vec_to_base(vec));

	fprintf(stderr, "metadata size: %zu\n", sizeof(cvector_metadata_t));
	fprintf(stderr, "start: %p\n", cvector_vec_to_base(vec));
	fprintf(stderr, "userptr: %p\n", vec);
	fprintf(stderr,
		"diff: %tu\n",
		(void *)vec - (void *)cvector_vec_to_base(vec));

	return 0;

}
```

</details>

**Comment by eteran** [comment]
OK, looks pretty good to me! Thanks!

**Pull request merged**